src/mlpack/tests/main_tests/random_forest_test.cpp

```diff
@@ -0,0 +1,219 @@
+  SetInputParam("input_model",
+                std::move(CLI::GetParam<RandomForestModel>("output_model")));
+
+  mlpackMain();
+
+  // Check that number of output points are equal to number of input points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("predictions").n_cols,
+                      testSize);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("probabilities").n_cols,
+                      testSize);
+
+  // Check number of output rows equals number of classes in case of
+  // probabilities and 1 for predicitions.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("predictions").n_rows,
+                      1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("probabilities").n_rows, 3);
+
+  // Check that initial predictions and predictions using saved model are same.
+  CheckMatrices(predictions, CLI::GetParam<arma::Row<size_t>>("predictions"));
+  CheckMatrices(probabilities, CLI::GetParam<arma::mat>("probabilities"));
+}
+
+/**
+ * Make sure number of trees specified is always a positive number.
+ */
+BOOST_AUTO_TEST_CASE(RandomForestNumOfTreesTest)
+{
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset vc2.csv!");
+
+  arma::Row<size_t> labels;
+  if (!data::Load("vc2_labels.txt", labels))
+    BOOST_FAIL("Cannot load labels for vc2_labels.txt");
+
+  SetInputParam("num_trees", (int) 0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure minimum leaf size specified is always a positive number.
+ */
+BOOST_AUTO_TEST_CASE(RandomForestMinimumLeafSizeTest)
+{
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset vc2.csv!");
+
+  arma::Row<size_t> labels;
+  if (!data::Load("vc2_labels.txt", labels))
+    BOOST_FAIL("Cannot load labels for vc2_labels.txt");
+
+  SetInputParam("minimum_leaf_size", (int) 0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure only one of training data or pre-trained model is passed.
+ */
+BOOST_AUTO_TEST_CASE(RandomForestTrainingVerTest)
+{
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset vc2.csv!");
+
+  arma::Row<size_t> labels;
+  if (!data::Load("vc2_labels.txt", labels))
+    BOOST_FAIL("Cannot load labels for vc2_labels.txt");
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+
+  mlpackMain();
+
+  // Input pre-trained model.
+  SetInputParam("input_model",
+                std::move(CLI::GetParam<RandomForestModel>("output_model")));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
```

Comment:
If you are willing to write some slightly more complicated tests, I think these could be added:

 * Ensure that training accuracy goes up when minimum_leaf_size is decreased: we could train on the vc2 dataset with minimum leaf size of 20, 10, and then 1, and ensure that the training error (which you'd have to calculate, but it's not too hard) is smallest for leaf size 1 and largest for leaf size 20.

 * Ensure that adding more trees decreases test set error (this one might need a little noise in the tolerance since it's not guaranteed).  You could build an RF on the vc2 set with 1 tree and with 5 trees, then ensure the test set accuracy of the 5-tree model is higher than the 1 tree model.

Let me know what you think.  Really here I am not aiming to test the random forest algorithm itself, but I do want to test that the leaf size and number of trees are properly getting passed to the random forest algorithm, and this is the best way I can think of to test that.  Maybe another idea is simply to inspect the RandomForestModel that is produced... that might be easier (I haven't fully thought it through yet).

Reply:
Done!